File: go-client/client_lib/cliente_sistema_arquivos.go
package clientlib

import (
	"context"
	"fmt"
	"log"
	"time"

	pb "ine5418/go-client/proto"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

type ClienteSistemaArquivos struct{
    conection *grpc.ClientConn 
    clientAPI pb.SistemaArquivosClient
    cache map[int]byte
    ultimoDescritor int
    versaoLocal int64
    cancelNotificacao context.CancelFunc

     retryExecutor *RetryExecutor
}

func NovoCliente(endereco string) (*ClienteSistemaArquivos, error) {
    ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
    defer cancel()

    conn, err := grpc.DialContext(ctx, endereco,
        grpc.WithTransportCredentials(insecure.NewCredentials()))

    if err != nil {
        fmt.Println("Erro ao criar conexao", err)
        return nil, err
    }
    
    cliente := &ClienteSistemaArquivos{
        conection: conn,
        clientAPI: pb.NewSistemaArquivosClient(conn),
        cache: make(map[int]byte),
        ultimoDescritor: -1,
        versaoLocal: 0,
 
        //retry com 3 tentativas
        retryExecutor: NewDefaultRetryExecutor(),
    }
    
    cliente.iniciarEscutaNotificacoes()
    
    return cliente, nil
}

func (c *ClienteSistemaArquivos) Abre(nome_arquivo string) int{
    resultado, err := c.executarComRetry(func() (int, error) {
        ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
        defer cancel()

        req := &pb.AbreRequest{
            NomeArquivo: nome_arquivo,
        }

        rply, err := c.clientAPI.Abre(ctx, req)
        if err != nil {
            return -1, err
        }

        if rply.GetStatus() < 0 {
            return -1, fmt.Errorf("Falha ao abrir arquivo. Status: %d", rply.GetStatus())
        }
        
        fmt.Printf("%s aberto, descritor: %d\n", nome_arquivo, rply.GetDescritor())
        return int(rply.GetDescritor()), nil
    })
    
    if err != nil {
        fmt.Println("Erro após retries:", err)
        return -1
    }
    return resultado
}

func (c *ClienteSistemaArquivos) Le(descritor int, posicao int, tamanho int) int {
   if c.ultimoDescritor != descritor {
        c.limparCache()
        c.ultimoDescritor = descritor
    }

    // Verifica cache primeiro (sem retry)
    cacheMiss := false
    conteudoCache := make([]byte, tamanho)
    for i := 0; i < tamanho; i++ {
        if val, ok := c.cache[posicao+i]; ok {
            conteudoCache[i] = val
        } else {
            cacheMiss = true
            break
        }
    }

    if !cacheMiss {
        fmt.Println("Conteudo lido da cache:", string(conteudoCache))
        return 0
    }

    resultado, err := c.executarComRetry(func() (int, error) {
        ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
        defer cancel()

        req := &pb.LeRequest{
            Descritor: int32(descritor),
            Posicao:   int32(posicao),
            Tamanho:   int32(tamanho),
        }

        rply, err := c.clientAPI.Le(ctx, req)
        if err != nil {
            return -1, err
        }

        if rply.GetStatus() < 0 {
            return -1, fmt.Errorf("Falha na leitura. Status: %d", rply.GetStatus())
        }

        fmt.Println("Conteudo lido:", string(rply.GetConteudoLer()))

        // Atualiza cache (sem retry)
        dados := rply.GetConteudoLer()
        for i := 0; i < len(dados); i++ {
            c.cache[posicao+i] = dados[i]
        }

        return int(rply.GetStatus()), nil
    })

    if err != nil {
        fmt.Println("Erro após retries:", err)
        return -1
    }
    return resultado
}

func (c *ClienteSistemaArquivos) Escreve(descritor int, posicao int, conteudo string) int {
    // Limpa cache local (sem retry, operação local)
    c.limparCache()
    
    if c.ultimoDescritor != descritor {
        c.limparCache()
        c.ultimoDescritor = descritor
    }

    resultado, err := c.executarComRetry(func() (int, error) {
        ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
        defer cancel()

        req := &pb.EscreveRequest{
            Descritor:        int32(descritor),
            Posicao:          int32(posicao),
            ConteudoEscrever: []byte(conteudo),
        }

        rply, err := c.clientAPI.Escreve(ctx, req)
        if err != nil {
            return -1, err
        }

        if rply.GetStatus() < 0 {
            return -1, fmt.Errorf("Falha na escrita. Status: %d", rply.GetStatus())
        }

        fmt.Println("Bytes escritos:", rply.GetBytesEscritos(), "Status: ", rply.GetStatus())

        // Atualiza cache local (sem retry, operação local)
        conteudoCache := []byte(conteudo)
        for i := 0; i < int(rply.GetBytesEscritos()); i++ {
            c.cache[posicao+i] = conteudoCache[i]
        }

        return int(rply.GetBytesEscritos()), nil
    })

    if err != nil {
        fmt.Println("Erro após retries:", err)
        return -1
    }
    return resultado
}

func (c *ClienteSistemaArquivos) Fecha(descritor int) int{
    ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
    defer cancel()

    req := &pb.FechaRequest{
        Descritor: int32(descritor),
    }

    rply, _ := c.clientAPI.Fecha(ctx, req)

    if rply.GetStatus() < 0{
        fmt.Println("Erro ao fechar o arquivo. Status:", rply.GetStatus())
    }
    return int(rply.GetStatus())
}

func (c *ClienteSistemaArquivos) limparCache() {
    for k := range c.cache {
        delete(c.cache, k)
    }
}
func (c *ClienteSistemaArquivos) executarComRetry(operacao RetryableOperation) (int, error) {
    return c.retryExecutor.ExecuteWithRetry(operacao)
}

func (c *ClienteSistemaArquivos) iniciarEscutaNotificacoes() {
    ctx, cancel := context.WithCancel(context.Background())
    c.cancelNotificacao = cancel
    
    req := &pb.NotificacaoRequest{
        ClientId: fmt.Sprintf("go-client-%d", time.Now().UnixNano()),
    }
    
    stream, err := c.clientAPI.RegistrarNotificacao(ctx, req)
    if err != nil {
        log.Printf("Erro ao registrar para notificações: %v", err)
        return
    }
    
    go func() {
        for {
            notificacao, err := stream.Recv()
            if err != nil {
                log.Printf("Erro ao receber notificação: %v", err)
                // tenta reconectar apos 5 segundos
                time.Sleep(5 * time.Second)
                c.iniciarEscutaNotificacoes()
                return
            }
            
            novaVersao := notificacao.GetVersaoGlobal()
            if novaVersao != c.versaoLocal {
                fmt.Printf("Notificação [SERVIDOR]: versão %d (era %d)\n", novaVersao, c.versaoLocal)
                c.versaoLocal = novaVersao
                c.limparCache() //invalida cache quando versão muda
            }
        }
    }()
    
    fmt.Println("Escuta de notificações iniciada")
}

func (c *ClienteSistemaArquivos) CloseConnection() error {
    // CANCELA A ESCUTA DE NOTIFICAÇÕES
    if c.cancelNotificacao != nil {
        c.cancelNotificacao()
    }
    return c.conection.Close()
}

File: java/src/main/java/com/mycompany/ine5418/client_lib/ClienteSistemaArquivos.java
package com.mycompany.ine5418.client_lib;

import com.mycompany.ine5418.client_lib.RetryExecutor;
import com.google.protobuf.ByteString;
import com.mycompany.ine5418.SistemaArquivosGrpc;
import com.mycompany.ine5418.SistemaArquivosProto;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.stub.StreamObserver;

import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CountDownLatch;

public class ClienteSistemaArquivos {
    private final ManagedChannel channel;
    private final SistemaArquivosGrpc.SistemaArquivosBlockingStub userStub;
    private final SistemaArquivosGrpc.SistemaArquivosStub asyncStub; // stub pra receber notificacao de cache invalida
    private final Map<Integer, Character> cache = new HashMap<>(); // implementacao da cache que mapeia uma posicao para um caractere
    private int ultimoDescritor = -1; // guarda o ultimo descritor utilizado, caso leia um arquivo de descritor diferente, substitui a cache pelo conteudo lido no servidor
    private long versaoLocal = 0; // guarda a versao do arquivo local para comparar com a do servidor quando for ler
    private final String clientId;
    private StreamObserver<SistemaArquivosProto.NotificacaoReply> notificacaoObserver;
    private final RetryExecutor retryExecutor;

    public ClienteSistemaArquivos(String host, int port) {
        this.channel = ManagedChannelBuilder.forAddress(host, port).usePlaintext().build();
        this.userStub = SistemaArquivosGrpc.newBlockingStub(this.channel);
        this.asyncStub = SistemaArquivosGrpc.newStub(this.channel);
        this.clientId = UUID.randomUUID().toString(); // ID único para este cliente

        this.retryExecutor = new RetryExecutor(3, 3000, 9000);

        iniciarEscutaNotificacoes();
    }
    public int Abre(String nome_arquivo) {
        return executarComRetry(() -> {
            SistemaArquivosProto.AbreRequest request = SistemaArquivosProto.AbreRequest.newBuilder()
                    .setNomeArquivo(nome_arquivo).build();
            SistemaArquivosProto.AbreReply reply = userStub.abre(request);

            if (reply.getStatus() >= 0) {
                System.out.println(nome_arquivo + " aberto, descritor: " + reply.getDescritor());
                return reply.getDescritor();
            }
            // ⭐ TRATA ERROS DE APLICAÇÃO COMO FALHAS (para fazer retry)
            throw new RuntimeException("Falha ao abrir arquivo. Status: " + reply.getStatus());
        });
    }

    public int Le(int descritor, int posicao, int tamanho){

        // se mudou o arquivo, reseta a cache (sem retry, operacao local)
        if (descritor != ultimoDescritor) {
            cache.clear();
            ultimoDescritor = descritor;
        }

        // percorre cada posicao desejada e verifica se esta presente na cache (sem retry, operacao local)
        boolean cacheMiss = false;
        char[] conteudo_cache = new char[tamanho];
        for (int i = 0; i<tamanho; i++){
            if (cache.get(posicao+i) == null) {
                cacheMiss = true;
                break;
            }
            conteudo_cache[i] = cache.get(posicao+i);
        }
        if (!cacheMiss){
            System.out.println("Conteudo lido da cache: "+ new String(conteudo_cache));
            return 0;
        }
        return executarComRetry(() -> {
            SistemaArquivosProto.LeRequest request = SistemaArquivosProto.LeRequest.newBuilder()
                    .setDescritor(descritor).setPosicao(posicao).setTamanho(tamanho).build();

            SistemaArquivosProto.LeReply reply = userStub.le(request);

            if (reply.getStatus() >= 0) {
                System.out.println("Conteudo lido: " + reply.getConteudoLer().toStringUtf8());

                byte[] dados = reply.getConteudoLer().toByteArray();
                for (int i = 0; i < dados.length; i++) {
                    cache.put(posicao + i, (char) dados[i]);
                }

                return reply.getStatus();
            } else {
                throw new RuntimeException("Falha na leitura. Status: " + reply.getStatus());
            }
        });
    }

    public int Escreve(int descritor, int posicao, String conteudo){
        cache.clear();

        // se mudou o arquivo, reseta a cache (sem retry, operacao local)
        if (descritor != ultimoDescritor) {
            cache.clear();
            ultimoDescritor = descritor;
        }

        // RETRY
        return executarComRetry(() -> {
            ByteString conteudoReq = ByteString.copyFrom(conteudo.getBytes(StandardCharsets.UTF_8));

            SistemaArquivosProto.EscreveRequest request = SistemaArquivosProto.EscreveRequest.newBuilder()
                    .setDescritor(descritor).setPosicao(posicao).setConteudoEscrever(conteudoReq).build();

            SistemaArquivosProto.EscreveReply reply = userStub.escreve(request);

            if (reply.getStatus() >= 0) {
                System.out.println("Bytes escritos: " + reply.getBytesEscritos());

                // Atualiza cache local (sem retry)
                char[] conteudo_cache = conteudo.toCharArray();
                for (int i = 0; i < reply.getBytesEscritos(); i++) {
                    cache.put(posicao + i, conteudo_cache[i]);
                }

                return reply.getBytesEscritos();
            } else {
                throw new RuntimeException("Falha na escrita. Status: " + reply.getStatus());
            }
        });
    }

    public int Fecha(int descritor){
        return executarComRetry(() -> {
            SistemaArquivosProto.FechaRequest request = SistemaArquivosProto.FechaRequest.newBuilder()
                    .setDescritor(descritor).build();
            SistemaArquivosProto.FechaReply reply = userStub.fecha(request);

            if (reply.getStatus() < 0) {
                throw new RuntimeException("Falha ao fechar arquivo. Status: " + reply.getStatus());
            }
            return reply.getStatus();
        });
    }

    // ================================================== METODOS DE COERENCIA DE CACHE =====================================================================================

    private void iniciarEscutaNotificacoes() {
        SistemaArquivosProto.NotificacaoRequest request =
                SistemaArquivosProto.NotificacaoRequest.newBuilder()
                        .setClientId(clientId)
                        .build();

        notificacaoObserver = new StreamObserver<SistemaArquivosProto.NotificacaoReply>() {
            @Override
            public void onNext(SistemaArquivosProto.NotificacaoReply notificacao) {
                long novaVersao = notificacao.getVersaoGlobal();
                if (novaVersao != versaoLocal) {
                    System.out.println("Notificação [SERVIDOR]: versão " + novaVersao + " (era " + versaoLocal + ")");
                    versaoLocal = novaVersao;
                    cache.clear(); // Invalida cache quando versão muda
                }
            }

            @Override
            public void onError(Throwable t) {
                System.err.println("Erro na escuta de notificações: " + t.getMessage());
            }

            @Override
            public void onCompleted() {
                System.out.println("Conexão de notificação fechada pelo servidor");
            }
        };

        asyncStub.registrarNotificacao(request, notificacaoObserver);
        System.out.println("Escuta de notificações iniciada para cliente: " + clientId);
    }
    // =================================================================================================================================

    private <T> T executarComRetry(RetryExecutor.RetryableOperation<T> operacao) {
        return retryExecutor.executeWithRetry(operacao);
    }




}